---
title: "BCB420 A3 - Dataset Pathway and Network Analysis"
author: "Yijia Chen"
date: "2022-04-05"
output:
  html_notebook:
    toc: true
bibliography: A3_citations.bib
csl: apa.csl
nocite: "@*"
---

```{r, message = FALSE, warning = FALSE, include = FALSE}
# Several R packages are required to run this notebook. They can be installed by running this snippet
# This snippet will not be included in the final report because the installation of EnhancedVolcano will always
# generate output that cannot be suppressed by message = FALSE or warning = FALSE

if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!requireNamespace("GEOquery", quietly = TRUE)) {
  BioCManager::install("GEOquery")
}
if (!requireNamespace("knitr", quietly = TRUE)) {
  install.packages("knitr")
}
if (!requireNamespace("edgeR", quietly = TRUE)) {
  BioCManager::install("edgeR")
}
if (!requireNamespace("ComplexHeatmap", quietly = TRUE)) {
  BiocManager::install("ComplexHeatmap")
}
if (!requireNamespace("circlize", quietly = TRUE)) {
  install.packages("circlize")
}
if (!requireNamespace("EnhancedVolcano", quietly = TRUE)) {
  BiocManager::install("EnhancedVolcano", update = FALSE, ask = FALSE)
}
library(EnhancedVolcano)
```

## 1. Introduction

This notebook contains a pathway and network analysis for the set of ranked genes obtained from the GEO dataset
[GSE157194](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157194), associated with the publication "Atopic
dermatitis displays stable and dynamic skin transcriptome signatures" [@mobus]. The goal of the study was to understand
the effects of the drugs dupilumab and cyclosporine on atopic dermatitis (AD) by conducting a gene expression study
using mRNA-seq data taken from biopsies of human patients.

The raw count matrix was first downloaded via GEOquery [@geoquery], and then the data was cleaned to confirm that no
duplicates were present. Next, the data was filtered by gene, removing those without at least one read per million in
each sample, and also by patient, removing those with incomplete data. Ensembl IDs were mapped to HUGO gene symbols
using biomaRt [@biomart], and genes that were unable to be converted (e.g., novel genes) were removed. To normalize the
data, trimmed mean of m-values was used [@edger], since the assumption was that most genes were not differentially
expressed.

The list of genes was used to perform a differential gene expression analysis using a quasi-likelihood model [@edger],
and the resulting p-values for each gene were converted to rank (using the formula
$\text{-}\hspace{-0.6mm}\log_{10}(p\text{-}value) \times \text{sign}(logFC)$). The resulting set of ranked genes was
sorted by descending rank and written to the file rankedGeneList.txt, which will be the starting point for GSEA.

## 2. Non-Thresholded Gene Set Enrichment Analysis


## 3. Visualization of Gene Set Enrichment Analysis (Cytoscape)


## 4. Interpretation


## 5. Step 3


## 6. References



















































